reviews/review-for-contributors.md

```diff
@@ -0,0 +1,45 @@
+# Mbed TLS Review Guidelines for Contributors
+
+## Introduction
+
+This document describes the process of having your pull request (PR) reviewed.
+
+## Goals of the review process
+
+The review process is there to ensure the quality (security, maintainability, etc.) of the code base. In order to be merged, a PR needs to pass the CI and be approved by at least two trusted reviewers; a trusted gatekeeper will then check everything's in order and press the merge button.
+
+See the [guideline for reviewers](<review_guidelines.md>) for a list of things reviewers will check.
+
+## General considerations
+
+As a project, we find reviewer bandwidth to be our most common bottleneck. As a consequence, the review process should be optimised for reviewers.
+
+Reviewers will generally assign themselves on a PR when there are available to review it, or when they commit to reviewing it later.
+
+Large contributions (say, more than 500 lines) should be discussed in advance on the mailing list, and as much as possible broken down into a series of smaller PRs.
+
+## Use of GitHub
+
+We use GitHub reviewing facilities; these have a few limitations which impact our review process (see below).
+
+## Reworking your PR in response to review comments
+
+* Please address feedback by adding commits to your PR, not by rewriting existing commits. Do so using atomic commits with descriptive messages. Avoid commits with titles like "address review feedback" that contains several unrelated changes.
+* Once you have addressed one point of feedback, please leave the GitHub discussion unresolved: the person who raise the point should mark it as resolved if they are satisfied with the resolution. This allows reviewers to easily track if all of their feedback as been addressed.
+  * If you want a way to mark which points you think you addressed, we tend to do that by reacting with the `:rocket:` emoji to the original comment. (We also use the thumbs up emoji as a way of saying we agree with the comment and intend to address it.) Note that this is purely optional.
+* Please avoid force-pushing once review has started, except in the circumstances listed below; this tends to disrupt incremental review and is not supported well enough by GitHub's reviewing tools. In particular, never force-push several times in a row; if you did, GitHub would to merge the events in the PR's timeline, showing only details of the last force-push, making it impossible to find again which commit was last reviewed. Exceptions:
```

Comment:
> GitHub would to merge the events in the PR's timeline, showing only details of the last force-push

That's not completely true, and there have been changes in the amount of information GitHub changes. As far as I remember, it's always been the case that the web interface (but not the API :frowning_face:) makes the history's history accessible if you dig enough, provided that some non-push even happens between two successive force-pushes (where “between” might require a few minutes on either side due to internal caching).

> never force-push several times in a row

I don't think we should forbid it, and in practice we do it often when the circumstances are right. For example:

1. Oh, there's a CI failure for something I can't test locally. Ok, I think this'll fix it. &lt;hackity&gt; &lt;push&gt; &lt;wait&gt;
2. Oh no, the fix didn't quite work right. &lt;hackity&gt; &lt;force-push&gt; &lt;wait&gt;
3. Repeat until CI passes.

I think force-pushing is generally ok _as long as nobody has responded (commented, or starting a review) to the commit you'll be overriding_, and as long as you're not squashing away useful insights about the history of the code.

(Case in point about “as long as nobody has responded”: GitHub prevented me from posting this comment when I was done writing it because you just force-pushed to rewrite the commit message…)

Reply:
> That's not completely true,

Agree, it won't happen all the time, but that fact that it can happen in some circumstances is annoying enough.

> I don't think we should forbid it

Agreed, again I'll adjust the strength of the wording.